### PR TITLE
Probably a much faster fuzzy matcher

### PR DESCRIPTION
**Pull request by BurningLutz**
Instead of regexp, I use two cursors to determine whether string B is a non-continuous sub string of String A.

### CODE REVIEW

**Comment by Shougo** [comment]
It is not so bad...
Please profile the improvement if it is much faster.
And why you have removed the binary search feature?

File: rplugin/python3/deoplete/filter/matcher_fuzzy.py [hunk 1]
```diff
@@ -18,24 +16,18 @@ def __init__(self, vim):
...
-        if context['is_sorted']:
-            begin = binary_search_begin(
-                context['candidates'], complete_str[0])
-            end = binary_search_end(
-                context['candidates'], complete_str[0])
-            if begin < 0 or end < 0:
-                return []
-            candidates = context['candidates'][begin:end+1]
-        else:
-            candidates = context['candidates']
```
Comment (Shougo) [inline rplugin/python3/deoplete/filter/matcher_fuzzy.py]: Why you have removed the binary search feature?
Please describe it.

Comment (BurningLutz) [reply rplugin/python3/deoplete/filter/matcher_fuzzy.py]: Hmmm...I'm not sure what is the binary search for. I thought it's for performance purpose, isn't it?

And I thought there is a bug of the binary search in the current fuzzy matcher. It doesn't reflect to the `ignorecase` option correctly.
Thus, the test case below will fail:
```python
    ctx = {
        'complete_str': 'fobr',
        'ignorecase': True,
        'camelcase': False,
        'is_sorted': False,
        'candidates': [
            { 'word': 'FooBar' },
            { 'word': 'foobar' },
            { 'word': 'aFooBar' },
            { 'word': 'fooBar' },
        ]
    }
    assert f.filter(ctx) == [
        { 'word': 'FooBar' },
        { 'word': 'foobar' },
        { 'word': 'fooBar' },
    ]
```
I will provide profiling results later, hmm.... maybe tomorrow.

**Comment by Shougo** [comment]
> Hmmm...I'm not sure what is the binary search for. I thought it's for performance purpose, isn't it?

Yes.

**Comment by Shougo** [comment]
> Thus, the test case below will fail:

In ignorecase, the order is undefined.

**Comment by Shougo** [comment]
> I will provide profiling results later, hmm.... maybe tomorrow

Please compare the huge candidate results.

**Comment by BurningLutz** [comment]
I'm afraid I can not provide profiling results because the f*cking GFW blocked gist.github.com and banned my VPN.

**Comment by Shougo** [comment]
> I'm afraid I can not provide profiling results today because the f*cking GFW blocked gist.github.com and banned my VPN.

Oh,...

**Comment by BurningLutz** [comment]
@Shougo 
I think I've made a mistake when doing profiling before I wrote the new fuzzy matcher.I tried to call `filter` many times to test its performance but it's not the case.

The real situation is that, `filter` is not called frequently so the most performance consuming `re.compile` affects performance little. And when doing candidates filtering, my `fuzzy_match` is slower than `p.match` at the most of time.

So I decide to close this PR.

**Comment by Shougo** [comment]
OK.

But can you update the tests?
Your tests method seems better.

**Comment by BurningLutz** [comment]
@Shougo 
OK. I'll do that.

**Pull request closed without merging**